iarc_roombafilter/scripts/f_ros.py

```diff
@@ -0,0 +1,170 @@
+
+        try:
+            # TODO : kind-of ugly hack to pull stamp into most recent available tf
+            # better way?
+            msg.fov_center.header.stamp = rospy.Time(0)
+            pos = self._tf.transformPoint('map', msg.fov_center)
+        except tf.Exception as e:
+            return
+
+        obs_ar = CircularObservation(
+                pos.point.x,
+                pos.point.y,
+                msg.fov_radius
+                )
+
+        # format observation data ...
+        obs = []
+        for r in msg.data:
+
+            #print r.visible_location.header.frame_id
+            #print r.visible_location.pose.pose.position.x
+
+            p = PoseStamped(r.visible_location.header, r.visible_location.pose.pose)
+            p = self._tf.transformPose('map', p).pose
+            p, q = p.position, p.orientation
+            h = tf.transformations.euler_from_quaternion([q.x,q.y,q.z,q.w])[2]
+
+            pose = np.asarray([p.x, p.y, h])
+            # note : no velocity information is available.
+
+            # roomba type handling
+            r_type = cfg.T_NULL
+            if r.type != Roomba.UNKNOWN:
+                if r.type == Roomba.OBSTACLE:
+                    r_type = cfg.T_OBST
+                else:
+                    r_type = cfg.T_RED if r.type == Roomba.RED else cfg.T_GREEN
+
+            o = ObservationParticle(
+                    pose=pose,
+                    t=r_type,
+                    )
+            obs.append(o)
+
+        # TODO(yoonyoungcho) does filterpy support retrodictive updates?
+        # TODO(yoonyoungcho) t = t0+dt or t1-dt?
+        #t = msg.header.stamp.to_sec()
+        t = rospy.Time.now().to_sec()
+        dt = t - self._t
+        if dt > 0:
+            self._mgr.step(obs, t, dt, obs_ar)
+            self._t = t
+
+    def publish(self):
+        now = rospy.Time.now()
+        cov = np.eye(6) * 1e-6 # TODO : do real covs
+        cov = cov.flatten()
+        msgs = []
+
+        #try:
+        #    print self._mgr.estimates()[0]._pose
+        #except Exception as e:
+        #    pass
+
+        for e in self._mgr.estimates():
+            (x,y,t,v,w) = e._pose
+            #print x, v
+            s,c = np.sin(t/2.0), np.cos(t/2.0)
+            p = Point(x,y,0)
+            q = Quaternion(0,0,s,c)
+            loc = PoseWithCovarianceStamped(
+                    Header(stamp=now, frame_id="map"),
+                    PoseWithCovariance(
+                        Pose(p,q),
+                        cov
+                    ))
+            msg = Roomba(now, "map", Roomba.UNKNOWN, loc)
+            msgs.append(msg)
+        msg = RoombaList(Header(stamp=now,frame_id="map"), msgs)
+        #print len(msgs)
+        self._pub.publish(msg)
+
+    def run(self):
+        rate = rospy.Rate(self._rate)
+        self._t = rospy.Time.now().to_sec()
+        while not rospy.is_shutdown():
+            #self._t = rospy.Time.now().to_sec()
+            self.obs_proc()
```

Comment:
Is there a reason this is running at a fixed rate, rather than just being part of the callback when a message comes in?

Reply:
to decouple observation processing (which takes time), so that older observations do not get processed at all.